zebra-network/src/peer_set/initialize.rs

```diff
@@ -358,26 +360,34 @@ where
 /// `peerset_initial_target_size`.
 ///
 /// The result is randomly chosen entries from the provided set of addresses.
-async fn limit_initial_peers(config: &Config) -> HashSet<SocketAddr> {
-    let initial_peers = config.initial_peers().await;
-    let initial_peer_count = initial_peers.len();
+async fn limit_initial_peers(
+    config: &Config,
+    mut timestamp_collector: mpsc::Sender<MetaAddrChange>,
+) -> HashSet<SocketAddr> {
+    let all_peers = config.initial_peers().await;
+    let peers_count = all_peers.len();
 
     // Limit the number of initial peers to `config.peerset_initial_target_size`
-    if initial_peer_count > config.peerset_initial_target_size {
+    if peers_count > config.peerset_initial_target_size {
         info!(
             "Limiting the initial peers list from {} to {}",
-            initial_peer_count, config.peerset_initial_target_size
+            peers_count, config.peerset_initial_target_size
         );
     }
 
-    let initial_peers_vect: Vec<SocketAddr> = initial_peers.iter().copied().collect();
+    // Split all the peers into the `initial_peers` that will be returned and
+    // `unused_peers` that will be sent to the address book.
+    let mut all_peers: Vec<SocketAddr> = all_peers.iter().copied().collect();
+    let (initial_peers, unused_peers) =
+        all_peers.partial_shuffle(&mut rand::thread_rng(), config.peerset_initial_target_size);
 
-    // TODO: add unused peers to the AddressBook (#2931)
-    //       https://docs.rs/rand/0.8.4/rand/seq/trait.SliceRandom.html#tymethod.partial_shuffle
-    initial_peers_vect
-        .choose_multiple(&mut rand::thread_rng(), config.peerset_initial_target_size)
-        .copied()
-        .collect()
+    // Send the unused peers to the address book.
+    for peer in unused_peers {
+        let peer_addr = MetaAddr::new_peer(*peer);
+        let _ = timestamp_collector.send(peer_addr).await;
```

Comment:
Documentation: We added an extra `await` on the `AddressBook` thread mutex, but it shouldn't deadlock.

Do you think we should add a test to make sure?

```suggestion
        // `send` only waits when the channel is full.
        // The timestamp collector is a separate task, so we will only wait for a short time.
        let _ = timestamp_collector.send(peer_addr).await;
```

Reply:
@teor2345 should the test just check if the `await` doesn't take more than a second, for example?

Reply:
That sounds good!

Here's what I'd suggest:
1. Create a `Config` with enough seed peers to fill up the initial target peer list, fill up the channel, and still have extra peers that will have to wait to send
2. Run `zebra_network::init()` with that config
3. Check that it doesn't deadlock

You'll probably want a utility method that calls `zebra_network::init()`:
https://github.com/ZcashFoundation/zebra/blob/9963471b7c13baf3e45f980c4ad13e0efe622e16/zebra-network/src/peer_set/initialize/tests/vectors.rs#L1198-L1200

With a config with lots of fake peers:
https://github.com/ZcashFoundation/zebra/blob/9963471b7c13baf3e45f980c4ad13e0efe622e16/zebra-network/src/peer_set/initialize/tests/vectors.rs#L1456-L1468

If you do get a deadlock, try inserting a yield inside the send loop:
https://docs.rs/tokio/1.13.0/tokio/task/fn.yield_now.html

Reply:
@jvff would you mind reviewing this test when it gets done?

Any test that makes sure we don't deadlock on channel overflow is fine. It doesn't need to do exactly what I suggested.

Reply:
I wrote the test here https://github.com/ZcashFoundation/zebra/pull/2974/commits/ee0574aa039fa12d2f8cf5b18fddd93b3a75f513.